[AIE2P] Optimize G_PHI regbankselect for fifo/acc

andcarminati opened the pull request:
Observation: this PR should include more tests after merging more fifo intrinsics.

andcarminati committed:
[AIE2P] Optimize regbankselect for fifo/acc

For G_IMPLICIT_DEF.
Looking through PHI nodes.

martien-de-jong commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/regbankselect-phi.mir:
```diff
@@ -0,0 +1,206 @@
+
+# Test 3:  The instruction  %19:_(p0) = G_PHI %0(p0), %bb.0, %27(p0), %bb.1
+# should not be mapped to accregbank. The result of the PHI node is used by an instruction
+# that is considered an `accululator-use instruction` but the considered result is just a scalar
```
nit accumulator

andcarminati replied:
Oh, did I really write this? Good catch!

gbossu commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/regbankselect-phi.mir:
```diff
@@ -0,0 +1,206 @@
+  bb.2:
+    PseudoRET implicit $lr
+
+...
```
Nit: Maybe add a test where a PHI is used both as a PHI and an ACC. I think we will favour the FIFO mapping at this point.

andcarminati replied:
Do you mean by both FIFO and ACC? Yes, we should favor the FIFO because we check it first. I hope we will not face this situation, according to the benchmarks I saw, what do you think?

khallouh replied:
Why should we favor the FIFO bank?

andcarminati replied:
I feel that this is a corner case situation. If we see this case in the future we can handle properly.

andcarminati commented on llvm/lib/Target/AIE/aie2p/AIE2PRegisterBankInfo.cpp:
```diff
@@ -966,6 +976,36 @@ AIE2PRegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
     }
     return AIEBaseRegisterBankInfo::getInstrMapping(MI);
   }
+  case TargetOpcode::G_PHI: {
```
This is a really special-case to handle. If we map to vector reg bank, we may block optimal SWP scheduling's after.  This is also a special case because one single PHI node may contain inputs mapped to fifo, accumulator and vector bank at the same time. The way it is mapped is also different, we handle only the output and inputs are related the inputs.

andcarminati replied:
Handled in a different way now.

khallouh commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/regbankselect-phi.mir:
```diff
@@ -0,0 +1,282 @@
+...
+
+# Test 3:  The instruction  %19:_(p0) = G_PHI %0(p0), %bb.0, %27(p0), %bb.1
+# should not be mapped to accregbank. The result of the PHI node is used by an instruction
```
Maybe add a similar test for a fifo incompatible type?

andcarminati replied:
Done! We have a implicit pointer definition in `test_fifo.ld.fill.implicit.ptr`.

khallouh commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/regbankselect-phi.mir:
```diff
@@ -0,0 +1,282 @@
+...
+
+# Test 4:  The instruction  %19:_(p0) = G_PHI %0(p0), %bb.0, %27(p0), %bb.1
+# is also used as PHI input.
```
Nit: do you mean? `%23:_(<32 x s32>) = G_PHI %66(<32 x s32>), %bb.0, %28(<32 x s32>), %bb.1`

andcarminati replied:
Oh, true.

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PRegisterBankInfo.cpp:
```diff
@@ -729,56 +733,62 @@ bool AIE2PRegisterBankInfo::hasFifoInput(const MachineInstr &MI,
+      if (DefReg.isPhysical())
+        continue;
+      if (isUseFifoInsn(MRI, TRI, DefReg, Depth + 1))
+        return true;
```
This is great that we now look into the real use. Can you add a test for this with bitcasts and copies? I don't think we have that for the fifo bank.

andcarminati replied:
Done.

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PRegisterBankInfo.cpp:
```diff
@@ -729,56 +733,62 @@ bool AIE2PRegisterBankInfo::hasFifoInput(const MachineInstr &MI,
+        continue;
+      if (isUseAccInsn(MRI, TRI, DefReg, Depth + 1))
+        return true;
+    } else if (MRI.getType(RegOp).getSizeInBits() == 2048 ||
```
This makes sense for 2048-bit vectors. Does it address a specific case that we did not handle or is this just to return earlier?

andcarminati replied:
In this case, I just kept the same check that we had before. I can investigate if we can remove.

khallouh replied:
After taking a look myself, it seems this just serves as an early return. Functionally, I don't think it's needed. We can keep it but in that case it can be moved up out of the loop since the size of `RegOp` doesn't depend on its uses (UseMI).

andcarminati replied:
I see the same, I will drop this check, in this way we can improve for a further refactor (integration with fifo handling code).

khallouh commented:
I left a couple of comments, where a couple of more tests would be nice. Otherwise, this looks good to me.

andcarminati commented:
> I left a couple of comments, where a couple of more tests would be nice. Otherwise, this looks good to me.

Hi @khallouh , thank you for the review. I hope I addressed all suggestions.

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PRegisterBankInfo.cpp:
```diff
@@ -745,56 +749,61 @@ bool AIE2PRegisterBankInfo::hasFifoInput(const MachineInstr &MI,
-                         usesAccReg(UseMI, MRI, TRI, TracedRegOp);
-                });
+    // skip copies
+    if (UseOpcode == TargetOpcode::G_BITCAST || IsCopyToVReg(UseMI) ||
```
Nit: ...copies, bitcasts and phis

khallouh reviewed (dismissed):
LGTM!

F-Stuckmann commented on llvm/lib/Target/AIE/aie2p/AIE2PRegisterBankInfo.cpp:
```diff
@@ -745,56 +749,61 @@ bool AIE2PRegisterBankInfo::hasFifoInput(const MachineInstr &MI,
+      Register DefReg = UseMI.getOperand(0).getReg();
+      if (DefReg.isPhysical())
+        continue;
+      if (isUseFifoInsn(MRI, TRI, DefReg, Depth + 1))
```
`isUseAccInsn` and `isUseFifoInsn` are very similar.
Could you pass the different functions as a function pointers? Then you don't duplicate code.

```

[isUseFifoInsn(const MachineRegisterInfo &MRI,
                                          const TargetRegisterInfo &TRI,
                                          const Register RegOp,
                                          unsigned Depth) const {
    return RecursiveOperandSearch(MRI, TRI, RegOp, depth, isUseFifoInsn, hasFifoInput) 
}

bool AIE2PRegisterBankInfo::isUseAccInsn(const MachineRegisterInfo &MRI,
                                         const TargetRegisterInfo &TRI,
                                         const Register &RegOp,
                                         unsigned Depth) const {
    return RecursiveOperandSearch(MRI, TRI, RegOp, depth, isUseAccInsn, usesAccReg) 
}
```

andcarminati replied:
Indeed they are asking for it. However we need a bit more discussions to harmonize with https://github.com/Xilinx/llvm-aie/pull/325.

andcarminati replied:
If possible, I would like to shift this refactor to the aforementioned PR.

F-Stuckmann replied:
that does sound like the correct place to do this

niwinanto commented on llvm/lib/Target/AIE/aie2p/AIE2PRegisterBankInfo.cpp:
```diff
@@ -745,56 +749,61 @@ bool AIE2PRegisterBankInfo::hasFifoInput(const MachineInstr &MI,
 
-  // Trace RegOp to its actual register, updating it if necessary
-  traceToActualReg(TracedRegOp);
+  for (auto &UseMI : MRI.use_nodbg_instructions(RegOp)) {
```
Do we really need to check for all the uses?

andcarminati replied:
Discussed offline. Now this cases is covered with the test case `test_fifo.ld.fill.store`.
